### PR TITLE
Start moving type definitions

### PR DESCRIPTION
**Pull request by canyon289**
Performing one to make sure this is correct before doing this a bunch more times.

1. Did I do this right for `theano.tensor.basic` for the one instance. If so I'll do the rest

2. For `theano.tensor.basic` the respective type file is `theano.tensor.type` but for `theano.scalar.basic` there is no such `type.py` in the submodule. Additionally the definitions there go through caching logic, so for some of the type definitions if they were moved I believe it would just create another dependency instead of removing one. Is my thinking correct?
https://github.com/pymc-devs/Theano-PyMC/blob/master/theano/scalar/basic.py#L730

Or perhaps maybe we just need to create a `theano.scalar.type` module and move all the associated code over so it mirrors the tensor module?

Fixes https://github.com/pymc-devs/Theano-PyMC/issues/53

### CODE REVIEW

**Comment by michaelosthege** [comment]
@canyon289 since #181 was merged, you can now merge `master` into your branch and use the notebook to identify targets for your refactoring.

File: tests/tensor/test_blas.py [hunk 1]
```diff
@@ -1722,7 +1723,7 @@ def setup_method(self):
 
         self.fa = tt.fscalar()
         self.da = tt.dscalar()
-        self.ca = tt.cscalar()
+        self.ca = theano.tensor.type.cscalar()
```
Comment (brandonwillard) [inline tests/tensor/test_blas.py:1726]: This should just be `tt.type.cscalar()`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Yeah, it looks like you have the right idea.  I wouldn't recommend doing this without the aid of some refactoring tools, though!

File: tests/tensor/test_blas.py (hunk 1 above)
Comment (canyon289) [reply tests/tensor/test_blas.py:1726]: It doesnt, but I added the import

**Comment by canyon289** [comment]
Running into a pretty nasty circular import. Ill dedicate a couple more hours to it this week to see if I can untangle things or make some progress.

@michaelosthege I tried using your notebook to figure out the loops but unfortunately because of the circular import itself I cant even load theano to inspect the import graph. Ill have to do this the plain old way through a debugger

**Comment by canyon289** [comment]
Moving these type definitions is causing numerous levels of circular import errors. Ill try for another day but if I dont see a light at the end of the tunnel will close this and let someone else give it a shot

**Comment by brandonwillard** [comment]
I noticed that you changed a lot of references from `tt.*` to `theano.tensor.type.*`; that will alter the import order and cause the problems you're seeing; however, since you've also added the relocated `theano.tensor.type` imports to `theano.tensor.__init__`, there's no need to change those references.

**Comment by canyon289** [comment]
The automated tool did that. Im writing a quick script that will go through
and replace with the tt style reference and Im hoping thatll fix the
problem! Fingers crossed

As always thanks for the feedback

**Comment by brandonwillard** [comment]
> Im writing a quick script that will go through and replace with the tt style reference

You can just remove the commit that makes those changes, no?

**Comment by canyon289** [comment]
> > Im writing a quick script that will go through and replace with the tt style reference
> 
> You can just remove the commit that makes those changes, no?

The commit that includes those changes is also the one that includes the type reference move since the refactoring tool does two things at once. At any rate got it done through some other means but it didnt help

**Comment by brandonwillard** [comment]
> The commit that includes those changes is also the one that includes the type reference move

That's fine; you can rebase before that point and apply only the changes you want to keep from that commit, then create a commit for that and remove the old commit.

**Comment by canyon289** [comment]
> > The commit that includes those changes is also the one that includes the type reference move
> 
> That's fine; you can rebase before that point and apply only the changes you want to keep from that commit, then create a commit for that and remove the old commit.

Yep, already done, commit history is cleaned up

**Comment by canyon289** [comment]
I'm relinquishing this issue ticket. After trying a couple approaches I can't see to get any better than guessing. Closing but leaving branch up as reference either to myself I have the time to reattempt this or someone else who wants to take a run at it

**Comment by brandonwillard** [comment]
No problem; this refactoring work can be a serious headache, and we definitely don't want people to get worn out by it.  Feel free to move on to another issue that's more interesting, or at least less circular-import-containing.

**Comment by canyon289** [comment]
Hey,
Thank you so much for the encouragement. It really means a lot and I truly appreciate it.

**Pull request closed without merging**